### PR TITLE
cuda ipc uct changes + gtest changes

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
@bureddy

All uct and gtest changes are here.

### CODE REVIEW

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c [hunk 1]
```diff
@@ -31,16 +33,165 @@ UCS_CLASS_DEFINE_NEW_FUNC(uct_cuda_ipc_ep_t, uct_ep_t, uct_iface_t*,
...
 UCS_CLASS_DEFINE_DELETE_FUNC(uct_cuda_ipc_ep_t, uct_ep_t);
 
-ucs_status_t uct_cuda_ipc_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov, size_t iovcnt,
-                                       uint64_t remote_addr, uct_rkey_t rkey,
-                                       uct_completion_t *comp)
+#define uct_cuda_ipc_trace_data(_remote_addr, _rkey, _fmt, ...)     \
+    ucs_trace_data(_fmt " to %"PRIx64"(%+ld)", ## __VA_ARGS__, (_remote_addr), \
+                   (_rkey))
+
+#define UCT_CUDA_IPC_ZERO_LENGTH_POST(len)                      \
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.c]: used only at one. may we can remove macro and directly use it

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c [hunk 2]
```diff
@@ -31,16 +33,165 @@ UCS_CLASS_DEFINE_NEW_FUNC(uct_cuda_ipc_ep_t, uct_ep_t, uct_iface_t*,
...
-                                       uct_completion_t *comp)
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_cuda_ipc_post_cuda_async_copy(uct_ep_h tl_ep, uint64_t remote_addr, const uct_iov_t *iov,
+                                  uct_rkey_t rkey, uct_completion_t *comp,
+                                  int direction)
 {
-    return UCS_OK;
+    uct_cuda_ipc_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_cuda_ipc_iface_t);
+    uct_cuda_ipc_ep_t *ep = ucs_derived_of(tl_ep, uct_cuda_ipc_ep_t);
+    uct_cuda_ipc_key_t *key = (uct_cuda_ipc_key_t *) rkey;
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.c]: sorry, i was wrong on = alignment. can you align all 3 assignments?

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c [hunk 3]
```diff
@@ -31,16 +33,165 @@ UCS_CLASS_DEFINE_NEW_FUNC(uct_cuda_ipc_ep_t, uct_ep_t, uct_iface_t*,
...
+    ucs_status_t status;
+    uct_cuda_ipc_event_desc_t *cuda_ipc_event;
+    CUstream stream;
+    ucs_queue_head_t *outstanding_queue;
+    void *mapped_rem_addr  = NULL;
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.c]: do we need to assign here? if yes, can move it up

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.c [hunk 4]
```diff
@@ -31,16 +33,165 @@ UCS_CLASS_DEFINE_NEW_FUNC(uct_cuda_ipc_ep_t, uct_ep_t, uct_iface_t*,
...
+    if (0 == iface->streams_initialized) {
+        status = uct_cuda_ipc_iface_init_streams(iface);
+        if (UCS_OK != status) {
+            return status;
+        }
+    }
+
+    stream = iface->stream_d2d[key->dev_num];
+    outstanding_queue = &iface->outstanding_d2d_event_q;
+    cuda_ipc_event = ucs_mpool_get(&iface->event_desc);
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.c]: align =

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.h [hunk 5]
```diff
@@ -1,35 +1,59 @@
...
 
 #include <uct/api/uct.h>
 #include <uct/base/uct_iface.h>
 #include <ucs/type/class.h>
+#include "cuda_ipc_md.h"
+
+
+#define UCT_CUDA_IPC_HASH_SIZE 256
+#define UCT_CUDA_IPC_PUT 0
+#define UCT_CUDA_IPC_GET 1
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.h]: move  UCT_CUDA_IPC_PUT/ UCT_CUDA_IPC_GET to cuda_ipc_ep.c as these used only in that file

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.h [hunk 6]
```diff
@@ -1,35 +1,59 @@
...
 
 typedef struct uct_cuda_ipc_ep_addr {
     int                ep_id;
 } uct_cuda_ipc_ep_addr_t;
 
+
 typedef struct uct_cuda_ipc_ep {
-    uct_base_ep_t           super;
+    uct_base_ep_t          super;
+    uct_cuda_ipc_rem_seg_t *rem_segments_hash[UCT_CUDA_IPC_HASH_SIZE];
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.h]: not used hash in this PR. remove rem_segments_hash and UCT_CUDA_IPC_HASH_SIZE

File: src/uct/cuda/cuda_ipc/cuda_ipc_md.c [hunk 7]
```diff
@@ -11,26 +12,26 @@
...
 {
-    md_attr->cap.flags         = UCT_MD_FLAG_REG |
-                                 UCT_MD_FLAG_NEED_RKEY;
+    md_attr->cap.flags         = UCT_MD_FLAG_REG | UCT_MD_FLAG_NEED_RKEY;
     md_attr->cap.reg_mem_types = UCS_BIT(UCT_MD_MEM_TYPE_CUDA);
     md_attr->cap.mem_type      = UCT_MD_MEM_TYPE_CUDA;
     md_attr->cap.max_alloc     = 0;
-    md_attr->cap.max_reg       = ULONG_MAX;
-    md_attr->rkey_packed_size  = 0;
+    md_attr->cap.max_reg       = UCT_CUDA_IPC_MAX_ALLOC_SZ; /*FIX THIS*/
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_md.c]: what need to fix?

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.c [hunk 8]
```diff
@@ -71,70 +69,226 @@ static ucs_status_t uct_cuda_ipc_iface_query(uct_iface_h iface,
                                           UCT_IFACE_FLAG_PENDING   |
                                           UCT_IFACE_FLAG_GET_ZCOPY |
                                           UCT_IFACE_FLAG_PUT_ZCOPY;
-
     iface_attr->cap.put.max_short       = 0;
     iface_attr->cap.put.max_bcopy       = 0;
     iface_attr->cap.put.min_zcopy       = 0;
-    iface_attr->cap.put.max_zcopy       = 0;
+    iface_attr->cap.put.max_zcopy       = UCT_CUDA_IPC_MAX_ALLOC_SZ;
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_iface.c]: why only 1<< 24?

File: src/uct/cuda/cuda_ipc/cuda_ipc_md.h [hunk 9]
```diff
@@ -1,29 +1,68 @@
...
 #define UCT_CUDA_IPC_MD_H
 
 #include <uct/base/uct_md.h>
+#include <uct/cuda/base/cuda_md.h>
+#include <uct/cuda/base/cuda_iface.h>
+#include <cuda.h>
+
+
+#define UCT_CUDA_IPC_MD_NAME      "cuda_ipc"
+#define CUDA_ERR_STRLEN           512
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_md.h]: not used

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.h [hunk 10]
```diff
@@ -13,17 +14,36 @@
...
+    ucs_mpool_t      event_desc;
+    ucs_queue_head_t outstanding_d2d_event_q;
+    int              device_count;
+    int              streams_initialized;
+    CUstream         stream_d2d[UCT_CUDA_IPC_MAX_PEERS];
+    int              p2p_map[UCT_CUDA_IPC_MAX_PEERS][UCT_CUDA_IPC_MAX_PEERS];
+    struct {
+        unsigned     max_poll;
+    } config;
 } uct_cuda_ipc_iface_t;
```
Comment (bureddy) [inline src/uct/cuda/cuda_ipc/cuda_ipc_iface.h]: add simple description for struct members

**Comment by bureddy** [comment]
@Akshay-Venkatesh  I'm yet look for supporting both cuda-ipc and cuda_copy.
Meantime, after fixing current review comments, can you make a PR to UCX public branch and  tag me there. we will see if there is coverity and other issues.   in the PR heading add "WIP- do not review" to prevent others to review now. what do you think?

**Comment by Akshay-Venkatesh** [comment]
I'll do that, Deva. Thanks for the review.

**Pull request merged**